client/core/bookie.go

```diff
@@ -395,6 +398,210 @@ func handleBookOrderMsg(_ *Core, dc *dexConnection, msg *msgjson.Message) error
+	}
+}
+
+// setMarketFinalEpoch revises the FinalEpoch and Persist fields of the named
+// market in the stored ConfigResponse.
+func (dc *dexConnection) setMarketFinalEpoch(name string, finalEpoch uint64, persist bool) {
+	dc.cfgMtx.Lock()
+	defer dc.cfgMtx.Unlock()
+	mkt := dc.marketConfig(name)
+	if mkt == nil {
+		return
+	}
+	mkt.FinalEpoch = finalEpoch
+	mkt.Persist = &persist
+}
+
+// handleTradeSuspensionMsg is called when a trade suspension notification is
+// received. This message may come in advance of suspension, in which case it
+// has a SuspendTime set, or at the time of suspension if subscribed to the
+// order book, in which case it has a Seq value set.
+func handleTradeSuspensionMsg(c *Core, dc *dexConnection, msg *msgjson.Message) error {
+	var sp msgjson.TradeSuspension
+	err := msg.Unmarshal(&sp)
+	if err != nil {
+		return fmt.Errorf("trade suspension unmarshal error: %v", err)
+	}
+
+	// Ensure the provided market exists for the dex.
+	mkt := dc.market(sp.MarketID)
+	if mkt == nil {
+		return fmt.Errorf("no market found with ID %s", sp.MarketID)
+	}
+
+	// Update the data in the stored ConfigResponse.
+	dc.setMarketFinalEpoch(sp.MarketID, sp.FinalEpoch, sp.Persist)
+
+	// SuspendTime == 0 means suspending now.
+	if sp.SuspendTime != 0 {
+		// This is just a warning about a scheduled suspension.
+		suspendTime := encode.UnixTimeMilli(int64(sp.SuspendTime))
+		detail := fmt.Sprintf("Market %s is now scheduled for suspension at %v", sp.MarketID, suspendTime)
+		c.notify(newServerNotifyNote("market suspend scheduled", detail, db.WarningLevel))
+		return nil
+	}
+
+	detail := fmt.Sprintf("Trading for market %s at %s is now suspended.", sp.MarketID, dc.acct.host)
+	if !sp.Persist {
+		detail += " All booked orders are now PURGED."
+	}
+	c.notify(newServerNotifyNote("market suspended", detail, db.WarningLevel))
+
+	if sp.Persist {
+		// No book changes. Just wait for more order notes.
+		return nil
+	}
+
+	// Clear the book and unbook/revoke own orders.
+	dc.booksMtx.RLock()
+	defer dc.booksMtx.RUnlock()
+
+	book, ok := dc.books[sp.MarketID]
+	if !ok {
+		return fmt.Errorf("no order book found with market id '%v'", sp.MarketID)
+	}
+
+	err = book.Reset(&msgjson.OrderBook{
+		MarketID: sp.MarketID,
+		Seq:      sp.Seq,        // forces seq reset, but should be in seq with previous
+		Epoch:    sp.FinalEpoch, // unused?
+		// Orders is nil
+	})
+	// Return any non-nil error, but still revoke purged orders.
+
+	// Revoke all active orders of the suspended market for the dex.
+	log.Warnf("Revoking all active orders for market %v.", mkt.Name)
+	updatedAssets := make(assetMap)
+	dc.tradeMtx.RLock()
+	for _, tracker := range dc.trades {
+		if tracker.Order.Base() == mkt.BaseID && tracker.Order.Quote() == mkt.QuoteID &&
+			tracker.metaData.Host == dc.acct.host && tracker.metaData.Status == order.OrderStatusBooked {
+			// Locally revoke the purged book order.
+			tracker.revoke()
+			updatedAssets.count(tracker.fromAssetID)
```

Comment:
Here's a solution to your hiccup with the JS not updating the cancellation status text to "revoked". Instead of trying to catch the status change in the "revoke" order-type notification that contains the cancel `core.Order`, you should instead send an order-type notification here, with the trade, not the cancel. The **order** status should then update correctly.

Reply:
I basically mirrored the order revoke logic from `(trackedTrade).revokeMatch` that was used for orders revoked as part of revoked matches, but that must have been broken.  The approach there was to create a core.Order with TargetID set (`// the important part for the frontend`).  Unless I'm not seeing the difference, we should review that to ensure orders revoked when their match is revoked get the proper UI updates too.

Reply:
Given the revokeMatch->revoke path relying on this approach, I resolved it on the frontend, although perhaps more heavy handed than required.  I'd like to defer any more refactoring to this unless necessary.